### PR TITLE
chore(backend): Add golangci-lint and pre-commit config

### PR DESCRIPTION
**Pull request by gkcalat**
**Description of your changes:**

- Added a popular golang linter `golangci-lint`, which is widely used across multiple popular projects
- Added a few sentences on code styling and formatting as well as best practices for developers

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: .golangci.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
+linters-settings:
+  misspell:
+    locale: US
+linters:
+  disable-all: true
+  enable:
```
Comment (chensun) [inline .golangci.yaml:6]: The list seems quite extensive and I'm not sure if they all make sense.
How about we follow kubernetes config: either https://github.com/kubernetes/kubernetes/blob/master/hack/golangci.yaml or https://github.com/kubernetes/kubernetes/blob/master/hack/golangci-strict.yaml

Comment (gkcalat) [reply .golangci.yaml]: Thank you. This is a great pointer. I adopted style and metalinters from k8s. I removed `ginkgolinter` and `logcheck` as they are specific for their repository. Instead, I added `misspell` (checks for common English misspelled words in comments) and `gosimple` (one of the default linters that specializes in simplifying code). See mode details about each linter [here](https://golangci-lint.run/usage/linters/).

**Review by chensun — APPROVED** [review]
/lgtm
/approve

Thank you!

**Pull request merged**